src/pages/ContributionsPage/ContributionForm.tsx

```diff
@@ -0,0 +1,209 @@
+import { useEffect, useMemo, useState } from 'react';
+
+import { useMyUser } from 'features/auth/useLoginData';
+import { useForm, useController } from 'react-hook-form';
+import { useMutation, useQueryClient } from 'react-query';
+
+import { ACTIVITIES_QUERY_KEY } from '../../features/activities/ActivityList';
+import { FormInputField } from 'components';
+import { QUERY_KEY_ALLOCATE_CONTRIBUTIONS } from 'pages/GivePage/EpochStatementDrawer';
+import { useCircleIdParam } from 'routes/hooks';
+import { ContentHeader, Text, Box, Button, Flex, MarkdownPreview } from 'ui';
+import { SaveState } from 'ui/SavingIndicator';
+
+import { createContributionMutation } from './mutations';
+import type { CurrentContribution } from './types';
+import { getCurrentEpoch } from './util';
+
+const NEW_CONTRIBUTION_ID = 0;
+
+export const ContributionForm = () => {
+  const circleId = useCircleIdParam();
+  const currentUserId = useMyUser(circleId)?.id;
+
+  const [saveState, setSaveState] = useState<{ [key: number]: SaveState }>({});
+  const [currentContribution, setCurrentContribution] =
+    useState<CurrentContribution | null>(null);
+
+  const [showMarkdown, setShowMarkDown] = useState<boolean>(false);
+
+  const queryClient = useQueryClient();
+
+  const { control, resetField, setValue, setFocus } = useForm({ mode: 'all' });
+
+  useEffect(() => {
+    // once we become buffering, we need to schedule
+    // this protection of state change in useEffect allows us to fire this only once
+    // so requests don't stack up
+    if (saveState[currentContribution?.contribution.id] == 'buffering') {
+      updateSaveStateForContribution(
+        currentContribution?.contribution.id,
+        'scheduled'
+      );
+    }
+  }, [
+    currentContribution?.contribution.id,
+    saveState[currentContribution?.contribution.id],
+  ]);
+
+  const { field: descriptionField } = useController({
+    name: 'description',
+    control,
+  });
+
+  const { mutate: createContribution, reset: resetCreateMutation } =
+    useMutation(createContributionMutation, {
+      onSuccess: async newContribution => {
+        queryClient.invalidateQueries({
+          queryKey: [QUERY_KEY_ALLOCATE_CONTRIBUTIONS],
+        });
+        setTimeout(
+          () => queryClient.invalidateQueries(ACTIVITIES_QUERY_KEY),
+          1000
```

Comment:
@CryptoGraffe @wingfeatherwave these seems a bit hacky... wait 1 sec then refetch... might be worth thinking if we can find a more reliable solution pre-public launch.

Reply:
The way to do that would be to make the activity insert inline with contribution creation and return it when creating contribution